File: python/bank-account/bank_account.py
class BankAccount:
    def __init__(self):
        self.opened = False

    def get_balance(self):
        if self.opened == False:
            raise ValueError('account not open')

        return self.balance

    def open(self):
        if self.opened == True:
            raise ValueError('account already open')

        self.balance = 0
        self.opened = True

    def deposit(self, amount):
        if self.opened == False:
            raise ValueError('account not open')
        if amount <= 0:
            raise ValueError('amount must be greater than 0')

        self.balance += amount

    def withdraw(self, amount):
        if self.opened == False:
            raise ValueError('account not open')
        if amount > self.balance:
            raise ValueError('amount must be less than balance')
        if amount <= 0:
            raise ValueError('amount must be greater than 0')

        self.balance -= amount

    def close(self):
        if self.opened == False:
            raise ValueError('account not open')

        self.opened = False

File: javascript/bank-account/bank-account.js
//
// This is only a SKELETON file for the 'Bank Account' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class BankAccount {
  constructor() {
    this.opened = false
  }

  open() {
    if (this.opened === true) {
      throw new ValueError('Account already open')
    }

    this.money = 0
    this.opened = true
  }

  close() {
    if (this.opened === false) {
      throw new ValueError('Account not open')
    }

    this.opened = false
  }

  deposit(amount) {
    if (this.opened === false) {
      throw new ValueError('Account not open')
    }

    if (amount <= 0) {
      throw new ValueError('Cannot deposit less than 1')
    }

    this.money += amount
  }

  withdraw(amount) {
    if (this.opened === false) {
      throw new ValueError('Account not open')
    }

    if (amount <= 0) {
      throw new ValueError('Cannot withdraw less than 1')
    }

    if (amount > this.money) {
      throw new ValueError('Cannot withdraw more than balance')
    }

    this.money -= amount
  }

  get balance() {
    if (this.opened === false) {
      throw new ValueError('Account not open')
    }

    return this.money
  }
}

export class ValueError extends Error {
  constructor() {
    super('Bank account error');
  }
}
